packages/components/tree/src/nested-tree-data-source.ts

```diff
@@ -0,0 +1,139 @@
+import {
+  TreeDataSource,
+  type TreeDataSourceMapping,
+  type TreeDataSourceNode,
+  type TreeDataSourceOptions
+} from './tree-data-source.js';
+
+export interface NestedTreeDataSourceMapping<T> extends TreeDataSourceMapping<T> {
+  getChildren(item: T): T[] | Promise<T[]> | undefined;
+}
+
+export interface NestedTreeDataSourceOptions<T> extends NestedTreeDataSourceMapping<T> {
+  loadChildren?(node: T): Promise<T[]>;
+  selects?: 'single' | 'multiple';
+}
+
+/**
+ * A tree model that represents a nested list of nodes.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export class NestedTreeDataSource<T = any> extends TreeDataSource<T> {
+  #mapping: NestedTreeDataSourceMapping<T>;
```

Comment:
Can you please add spacing between variables and maybe add descriptions as well?